cli/cmd/inject.go

```diff
@@ -278,71 +287,109 @@ func (options *injectOptions) fetchConfigsOrDefault() (*config.All, error) {
 	if len(options.ignoreInboundPorts) > 0 {
 		configs.Proxy.IgnoreInboundPorts = toPorts(options.ignoreInboundPorts)
+		overrideAnnotations[k8s.ProxyIgnoreInboundPortsAnnotation] = parsePorts(configs.Proxy.IgnoreInboundPorts)
 	}
 	if len(options.ignoreOutboundPorts) > 0 {
 		configs.Proxy.IgnoreOutboundPorts = toPorts(options.ignoreOutboundPorts)
+		overrideAnnotations[k8s.ProxyIgnoreOutboundPortsAnnotation] = parsePorts(configs.Proxy.IgnoreOutboundPorts)
 	}
 
 	if options.proxyAdminPort != 0 {
 		configs.Proxy.AdminPort = toPort(options.proxyAdminPort)
+		overrideAnnotations[k8s.ProxyAdminPortAnnotation] = parsePort(configs.Proxy.AdminPort)
 	}
 	if options.proxyControlPort != 0 {
 		configs.Proxy.ControlPort = toPort(options.proxyControlPort)
+		overrideAnnotations[k8s.ProxyControlPortAnnotation] = parsePort(configs.Proxy.ControlPort)
 	}
 	if options.proxyInboundPort != 0 {
 		configs.Proxy.InboundPort = toPort(options.proxyInboundPort)
+		overrideAnnotations[k8s.ProxyInboundPortAnnotation] = parsePort(configs.Proxy.InboundPort)
 	}
 	if options.proxyOutboundPort != 0 {
 		configs.Proxy.OutboundPort = toPort(options.proxyOutboundPort)
+		overrideAnnotations[k8s.ProxyOutboundPortAnnotation] = parsePort(configs.Proxy.OutboundPort)
 	}
 
-	if options.dockerRegistry != "" {
-		configs.Proxy.ProxyImage.ImageName = registryOverride(configs.Proxy.ProxyImage.ImageName, options.dockerRegistry)
-		configs.Proxy.ProxyInitImage.ImageName = registryOverride(configs.Proxy.ProxyInitImage.ImageName, options.dockerRegistry)
+	if options.proxyImage != "" {
+		configs.Proxy.ProxyImage.ImageName = options.proxyImage
+		if options.dockerRegistry != "" {
+			configs.Proxy.ProxyImage.ImageName = registryOverride(configs.Proxy.ProxyImage.ImageName, options.dockerRegistry)
+		}
+		overrideAnnotations[k8s.ProxyImageAnnotation] = configs.Proxy.ProxyImage.ImageName
+	}
+
+	if options.initImage != "" {
+		configs.Proxy.ProxyInitImage.ImageName = options.initImage
+		if options.dockerRegistry != "" {
+			configs.Proxy.ProxyInitImage.ImageName = registryOverride(configs.Proxy.ProxyInitImage.ImageName, options.dockerRegistry)
+		}
+		overrideAnnotations[k8s.ProxyInitImageAnnotation] = configs.Proxy.ProxyInitImage.ImageName
 	}
 
 	if options.imagePullPolicy != "" {
 		configs.Proxy.ProxyImage.PullPolicy = options.imagePullPolicy
 		configs.Proxy.ProxyInitImage.PullPolicy = options.imagePullPolicy
+		overrideAnnotations[k8s.ProxyImagePullPolicyAnnotation] = options.imagePullPolicy
 	}
 
 	if options.proxyUID != 0 {
 		configs.Proxy.ProxyUid = options.proxyUID
+		overrideAnnotations[k8s.ProxyUIDAnnotation] = strconv.FormatInt(options.proxyUID, 10)
 	}
 
 	if options.proxyLogLevel != "" {
 		configs.Proxy.LogLevel = &config.LogLevel{Level: options.proxyLogLevel}
+		overrideAnnotations[k8s.ProxyLogLevelAnnotation] = options.proxyLogLevel
 	}
 
+	configs.Proxy.DisableExternalProfiles = options.disableExternalProfiles
```

Comment:
Track this even if it's `false`, so that the `LINKERD2_PROXY_DESTINATION_PROFILE_SUFFIXES` env var will be assign the correct value.

Reply:
Do you mind adding that comment in the code?

Reply:
Hmm, yeah, I'm a bit confused as to why it's necessary to set this if `options.disableExternalProfiles` is false (the default). I'd expect this field to only be set if `options.disableExternalProfiles` is true.

Reply:
note that there's an open P0 to change this

Reply:
Yep, #2534 is tracking that. But all the same, I think we should only set the annotation if the flag is set to the non-default value (whatever that may be).

Reply:
@klingerf This line only stores the value in the config struct, which is needed in the shared lib to decide on the value of `LINKERD2_PROXY_DESTINATION_PROFILE_SUFFIXES`. The annotation is added in the next line, only if the flag has a non-default value.